internal/contributebot/main.go

```diff
@@ -0,0 +1,167 @@
+// limitations under the License.
+
+// contributebot is a service for keeping the Go Cloud project tidy.
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/google/go-github/github"
+)
+
+const userAgent = "google/go-cloud Contribute Bot"
+
+type flagConfig struct {
+	project      string
+	subscription string
+	gitHubAppID  int64
+	keyPath      string
+}
+
+func main() {
+	addr := flag.String("listen", ":8080", "address to listen for health checks")
+	var cfg flagConfig
+	flag.StringVar(&cfg.project, "project", "", "GCP project for topic")
+	flag.StringVar(&cfg.subscription, "subscription", "contributebot-github-events", "subscription name inside project")
+	flag.Int64Var(&cfg.gitHubAppID, "github_app", 0, "GitHub application ID")
+	flag.StringVar(&cfg.keyPath, "github_key", "", "path to GitHub application private key")
+	flag.Parse()
+	if cfg.project == "" || cfg.gitHubAppID == 0 || cfg.keyPath == "" {
+		fmt.Fprintln(os.Stderr, "contributebot: must specify -project, -github_app, and -github_key")
+		os.Exit(2)
+	}
+
+	ctx := context.Background()
+	w, server, cleanup, err := setup(ctx, cfg)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer cleanup()
+	log.Printf("Serving health checks at %s", *addr)
+	go server.ListenAndServe(*addr, w)
+	log.Fatal(w.receive(ctx))
+}
+
+// worker contains the connections used by this server.
+type worker struct {
+	sub  *pubsub.Subscription
+	auth *gitHubAppAuth
+}
+
+// receive listens for events on its subscription and handles them.
+func (w *worker) receive(ctx context.Context) error {
+	return w.sub.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
+		id := msg.Attributes["X-GitHub-Delivery"]
+		eventType := msg.Attributes["X-GitHub-Event"]
+		if eventType == "integration_installation" || eventType == "integration_installation_repositories" {
+			// Deprecated event types. Ignore them in favor of supported ones.
+			log.Printf("Skipped event %s of deprecated type %s", id, eventType)
+			msg.Ack()
+			return
+		}
+		event, err := github.ParseWebHook(eventType, msg.Data)
+		if err != nil {
+			log.Printf("Parsing %s event %s: %v", eventType, id, err)
+			msg.Nack()
+			return
+		}
+		var handleErr error
+		switch event := event.(type) {
+		case *github.IssuesEvent:
+			handleErr = w.receiveIssueEvent(ctx, event)
+		case *github.PingEvent, *github.InstallationEvent, *github.PullRequestEvent, *github.CheckSuiteEvent:
+			// No-op.
+		default:
+			log.Printf("Unhandled webhook event type %s (%T) for %s", eventType, event, id)
+			msg.Nack()
+			return
+		}
+		if handleErr != nil {
+			log.Printf("Failed processing %s event %s: %v", eventType, id, handleErr)
+			msg.Nack()
+			return
+		}
+		msg.Ack()
+		log.Printf("Processed %s event %s", eventType, id)
+	})
+}
+
+func (w *worker) receiveIssueEvent(ctx context.Context, e *github.IssuesEvent) error {
+	const inProgressLabel = "in progress"
```

Comment:
Maybe put this in a func, e.g., "checkForClosedIssueWithInProgress" or something? Presumably we will end up adding more checks.

Reply:
Added a bit more structure, since there's somewhat of a two-phase test-and-run procedure. Let me know what you think.